sbncode/CAFMaker/FillReco.cxx

```diff
@@ -950,6 +970,59 @@ namespace caf
     srpfp.cnnscore.nclusters = cnnscore->nClusters;
   }
 
+  void FillPPFNuGraph(const std::vector<unsigned int> &sliceHitsMap,
+		      const std::vector<art::Ptr<anab::FeatureVector<1>>> &ngFilterResult,
+		      const std::vector<art::Ptr<anab::FeatureVector<5>>> &ngSemanticResult,
+		      const std::vector<art::Ptr<recob::Hit>> &pfpHits,
+		      caf::SRPFP& srpfp,
+		      bool allowEmpty)
+  {
+
+    auto arg_max = [](std::vector<float> const& vec) {
+      return std::distance(vec.begin(), std::max_element(vec.begin(), vec.end()));
+    };
+
+    // the nugraph elements are ordered the same as the sliceHitsMap
+    std::vector<size_t> mappedhits;
+    for (auto& hit : pfpHits) {
+      auto it = std::find(sliceHitsMap.begin(), sliceHitsMap.end(), hit.key());
+      if (it != sliceHitsMap.end()) {
+	size_t index = std::distance(sliceHitsMap.begin(), it);
+	mappedhits.push_back(index);
+      }
+    }
+
+    if (mappedhits.size()>0) {
+      std::vector<float> ng2sempfpcounts(5,0);
+      size_t ng2bkgpfpcount = 0;
+      for (size_t pos : mappedhits) {
+	auto scores = ngSemanticResult.at(pos);
+	std::vector<float> ng2semscores;
+	for (size_t i=0;i<scores->size();i++) ng2semscores.push_back(scores->at(i));
+	size_t sem_label = arg_max(ng2semscores);
+	ng2sempfpcounts[sem_label]++;
+	auto bkgscore = ngFilterResult.at(pos);
+	if (bkgscore->at(0)<0.5) ng2bkgpfpcount++;
```

Comment:
This seems fine, but I was thinking: is there an argument to be made that one should only fill the semantic label fractions for hits that are not tagged as background?

Reply:
Yes, that's a good point. Ideally this should be studied, to find what works best. In general, the semantic labels can be meaningful also when the filter does not pass (e.g. for cosmics). But since noise is a bigger issues than cosmics (unlike MicroBooNE), your suggestion makes sense and I will implement it

Reply:
yes, the current training is considering them signal but this is going to change soon!

Reply:
Ah then I guess I'm okay with either way, though I suppose in this updated schema the background fraction would still be meaningful to distinguish a cosmic PFP from others? And the semantic label fractions would help distinguish if there is some overlapping particle what it was, no?

Reply:
right. My point is that for ICARUS noise is likely more of an issue than cosmics, so the semantic labels for the background hits are more likely to be meaningless (as they are more likely to be noise and not real cosmic particles). So the new version makes more sense as we don't want to pollute the semantic counts with noise. But of course the background fraction is still useful to reject cosmics!